sw/device/silicon_creator/lib/drivers/usb.c

```diff
@@ -0,0 +1,463 @@
+}
+
+static void send_packet(uint8_t ep_index) {
+  usb_ep_info_t *endpoint = in_endpoints + ep_index;
+  size_t chunk = endpoint->max_packet_size < endpoint->transfer.len
+                     ? endpoint->max_packet_size
+                     : endpoint->transfer.len;
+  uint8_t buffer = buffer_pool_get();
+
+  if (chunk < endpoint->max_packet_size) {
+    // If the chunk is shorter than the endpoint size, then we can
+    // clear the ShortIn flag.
+    endpoint->transfer.flags &= ~(uint32_t)kUsbTransactionFlagsShortIn;
+  }
+  copy_to_buffer(buffer, (char *)endpoint->transfer.data, chunk);
+  endpoint->transfer.data += chunk;
+  endpoint->transfer.len -= chunk;
+
+  uint32_t val = 0;
+  val = bitfield_field32_write(val, USBDEV_CONFIGIN_0_BUFFER_0_FIELD, buffer);
+  val = bitfield_field32_write(val, USBDEV_CONFIGIN_0_SIZE_0_FIELD, chunk);
+
+  // Mark the packet as ready for transmission
+  val = bitfield_bit32_write(val, USBDEV_CONFIGIN_0_RDY_0_BIT, true);
+  abs_mmio_write32(
+      kBase + USBDEV_CONFIGIN_0_REG_OFFSET + ep_index * sizeof(uint32_t), val);
+}
+
+rom_error_t usb_ep_transfer(uint8_t ep, void *data, size_t len,
+                            usb_transaction_flags_t flags) {
+  uint8_t i = ep & kEpNumMask;
+  if (i >= USBDEV_PARAM_N_ENDPOINTS) {
+    return kErrorUsbBadEndpointNumber;
+  }
+  usb_ep_info_t *endpoint = ep & kUsbDirIn ? in_endpoints : out_endpoints;
+  endpoint += i;
+
+  if (endpoint->type & kUsbEpTypeControl && len > 0) {
+    // Transfers of more than length zero on a control endpoint require a
+    // zero-length transfer in the opposite direction to finish the transaction.
+    flags |= kUsbTransactionFlagsZlp;
+  }
+  endpoint->transfer.data = (char *)data;
+  endpoint->transfer.len = len;
+  endpoint->transfer.bytes_transfered = 0;
+  endpoint->transfer.flags = flags;
+  if (ep & kUsbDirIn) {
+    // IN transfer to host; send the first packet.
+    send_packet(i);
+  } else {
+    // OUT transfer from host; enable receiving OUT packets.
+    usbreg_bit(USBDEV_RXENABLE_OUT_REG_OFFSET, i, true);
+  }
+  return kErrorOk;
+}
+
+void handle_in(void) {
+  // Handle IN transactions: return sent buffers to the buffer pool and send the
+  // next packet in any multi-packet transfers.
+  uint32_t sent = abs_mmio_read32(kBase + USBDEV_IN_SENT_REG_OFFSET);
+  for (uint8_t ep = 0; sent && ep < USBDEV_PARAM_N_ENDPOINTS; ++ep) {
+    if ((sent & (1 << ep)) == 0) {
+      continue;
+    }
+    sent &= ~(1 << ep);
+    usb_ep_info_t *endpoint = in_endpoints + ep;
+    uint32_t reg = abs_mmio_read32(kBase + USBDEV_CONFIGIN_0_REG_OFFSET +
+                                   ep * sizeof(uint32_t));
+    uint8_t buffer =
+        (uint8_t)bitfield_field32_read(reg, USBDEV_CONFIGIN_0_BUFFER_0_FIELD);
+    buffer_pool_put(buffer);
+    abs_mmio_write32(
+        kBase + USBDEV_CONFIGIN_0_REG_OFFSET + ep * sizeof(uint32_t),
+        1 << USBDEV_CONFIGIN_0_PEND_0_BIT);
```

Comment:
If ever there is a day where this is part of an actual interrupt handler, you might want to put retiring the buffer _after_ the CSR write here. :smile:

Reply:
Fixed.